### PR TITLE
fix: instance type and architecture incompatiblity

### LINKED ISSUE #25: ARM64 v.s. x86_64
By default this module selects [`t4g.nano`](https://github.com/masterpointio/terraform-aws-ssm-agent/blob/5b6d9947bbffb2a9ced75186a6059fbe62208e67/variables.tf#L22) as the instance type which is arm64 based, but [in `data.tf`](https://github.com/masterpointio/terraform-aws-ssm-agent/blob/5b6d9947bbffb2a9ced75186a6059fbe62208e67/data.tf#L16), the filtering is based on x86 only which is causing an error when applying since (naturally) an x86 machine image does not run on arm64 machines. 

I don't know if there's some easy way to fix this, but I thought I'd mention it since the defaults seem to be broken at the moment?

The error I see is:

```
api error ValidationError: You must use a valid fully-formed launch template. The architecture 'arm64' of the specified instance type does not match the architecture 'x86_64' of the specified AMI. Specify an instance type and an AMI that have matching architectures, and try again. You can use 'describe-instance-types' or 'describe-images' to discover the architecture of the instance type or AMI."
```

### PR DESCRIPTION
**Pull request by oycyc**
## what
* Makes the AMI architecture value a variable so it can be passed in. Default set to arm64 to be compatible with t4g.nano default, which is cheaper and seemingly more efficient compared to t3.nano.
* Maybe @Gowiem or @gberenice , you two might have more context, but looking at it and reading through, there doesn't seem like anything that requires a strict x86_64 architecture.
* I've tested by using this module with the example with different combination, and it all spins up the resources smoothly.
* Also added validation to ensure we catch this error. 

## references
* Closes #25

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1,3 +1,25 @@
+locals {
+  # Validate that only 'arm64' architecture is used with 'g' processor instances to ensure compatibility.
+  # https://docs.aws.amazon.com/ec2/latest/instancetypes/instance-type-names.html
+  is_instance_compatible = (
```
Comment (gberenice) [inline main.tf:4]: `g` can be present in the instance family: `G – Graphics intensive`. This is obviously a weird choice for an SSM agent, but I think we still need to ensure we validate the processor family field.
According to the docs, this is a third position, and we can use the function `element(list, index)` to check this.
Thoughts on making this check stricter?

Comment (oycyc) [reply main.tf:4]: The `g` isn't representing the instance family, it's representing the processor family of that instance, and it represents `AWS Graviton processors` which is AWS' ARM based processors

Comment (oycyc) [reply main.tf:4]: Oh I see what you mean now, that the validation would be valid because it currently only checks for any 'g's. But having the stricter check will ensure that the processor family flag has a 'g'.

Yeah that makes sense then, definitely will make this check stricter.

Comment (oycyc) [reply main.tf:5]: Just added those stricter validation, and also tinkered with some native TF tests to cover those edge cases. 

That's a great catch about the instance family, thanks! Yeah it would be strange to use for an SSM agent, just looking at their lowest pricing. But always nice to have these validations in place. 

image

**Comment by oycyc** [comment]
Just added those stricter validation, and also tinkered with some native TF tests to cover those edge cases. 

That's a great catch about the instance family, thanks! Yeah it would be strange to use for an SSM agent, just looking at their lowest pricing. But always nice to have these validations in place. 

image

**Comment by oycyc** [comment]
Might be cool in the future if we start using tests more to have a GH actions that runs these automatically as a check on all our modules!

**Comment by gberenice** [comment]
> Might be cool in the future if we start using tests more to have a GH actions that runs these automatically as a check on all our modules!

Yep, agree. That's what we have in the roadmap 👍

File: main.tf (hunk 1 above)
Comment (gberenice) [reply main.tf:5]: Looks great! 🤩

**Review by gberenice — APPROVED** [review]
LGTM :shipit:
